### PR TITLE
APIM: Api / Product / VersionSet / Policy / PolicyFragment

### PR DESCRIPTION
**Pull request by ross-p-smith**
Relates to #2747 and builds on #3376

This is the last of my API Management additions and this should now give a bare minimum needed to deploy apis in a cloud native manner.

**Special notes for your reviewer**:
The Product/Subscription Integration Test is commented out - #3408
Implements #3379 

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains tests

### CODE REVIEW

**Comment by ross-p-smith** [comment]
@matthchr - does this implement what you need for #3379 ?

**Review by theunrepentantgeek — COMMENTED** [review]
Looks good - a couple of minor comments. I'll have another pass on my Monday.

File: v2/internal/controllers/crd_apimanagement_20220801_test.go [hunk 1]
```diff
@@ -24,33 +26,38 @@ func Test_ApiManagement_20220801_CRUD(t *testing.T) {
...
-			// Restore:        to.Ptr(true),
-		},
+			ObjectMeta: tc.MakeObjectMetaWithName(tc.Namer.GenerateName("apim")),
+			Spec: apim.Service_Spec{
+					Location:       tc.AzureRegion,
+					Owner:          testcommon.AsOwner(rg),
+					PublisherEmail: to.Ptr("ASO@testing.com"),
+					PublisherName:  to.Ptr("ASOTesting"),
+					Sku:            &sku,
+					Restore:        to.Ptr(true),
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply v2/internal/controllers/crd_apimanagement_20220801_test.go:54]: To clarify - if you think such a feature would be valuable, we can look at doing that as a separate PR. I don't want to hold this one up with additional complexity.

**Comment by tomkerkhove** [comment]
@ross-p-smith looks like this also supports policy fragment, mind adding it to the title?

File: v2/internal/controllers/crd_apimanagement_20220801_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply v2/internal/controllers/crd_apimanagement_20220801_test.go:54]: Great idea, I did see your PR there for a while and thought it was a great idea to add  #3433

File: v2/samples/apimanagement/v1api20220801/v1api20220801_service.yaml [hunk 2]
```diff
@@ -12,3 +12,4 @@ spec:
     name: aso-sample-rg
   publisherEmail: ASO@testing.com
   publisherName: ASOTesting
+  #restore: true # The service may have been soft-deleted previously, so we need to restore it
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply v2/samples/apimanagement/v1api20220801/v1api20220801_service.yaml]: Removed - (for reference #3433)

**Comment by ross-p-smith** [comment]
Hi @matthchr, @theunrepentantgeek 

```
../../api/apimanagement/customizations/subscription_extensions.go:11:2: missing go.sum entry for module providing package github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/apimanagement/armapimanagement (imported by github.com/Azure/azure-service-operator/v2/api/apimanagement/customizations); to add:
[asoctl:unit-tests-cover] 	go get github.com/Azure/azure-service-operator/v2/api/apimanagement/customizations@v2.2.0
```

I'm not a go person and struggling to know how to fix this

File: v2/internal/controllers/crd_apimanagement_20220801_test.go [hunk 3]
```diff
@@ -81,37 +88,86 @@ func Test_ApiManagement_20220801_CRUD(t *testing.T) {
...
+	// The subscription will have been created and it would have generated it's own
+	// PrimaryKey and SecondaryKey. Now let's see if we can set them from a Kubernetes secret.
+
+	// There should be no secrets at this point
+	secretList := &v1.SecretList{}
+	tc.ListResources(secretList, client.InNamespace(tc.Namespace))
+	tc.Expect(secretList.Items).To(HaveLen(0))
+
+	// Run sub-tests on subscription
+	tc.RunSubtests(
```
Comment (theunrepentantgeek) [inline v2/internal/controllers/crd_apimanagement_20220801_test.go:144]: If these tests must be run in sequence (not in parallel), best to add a comment to that effect:

```suggestion
	// Run sub-tests serially on subscription
	tc.RunSubtests(
```

We're constantly trying to keep our CI runs as quick as practical, and a common maintenance task is to make tests run in parallel where possible.

**Review by theunrepentantgeek — APPROVED** [review]

**Comment by theunrepentantgeek** [comment]
I think your CI test failure in `test-generator` was a flake (I have work underway to improve the reliability of our recorded tests), so I've triggered a re-run of that check. 

I've also triggered a full run of `integration-tests-fork` as well as the `Azure.azure-service-operator` check.

File: v2/internal/controllers/crd_apimanagement_20220801_test.go (hunk 3 above)
Comment (ross-p-smith) [reply v2/internal/controllers/crd_apimanagement_20220801_test.go]: Added a comment to all of the occurrences of this pattern

**Pull request merged**